engine/src/main/java/org/terasology/rendering/nui/widgets/types/builtin/util/FieldsWidgetBuilder.java

```diff
@@ -0,0 +1,190 @@
+import org.terasology.rendering.nui.UIWidget;
+import org.terasology.rendering.nui.WidgetUtil;
+import org.terasology.rendering.nui.databinding.Binding;
+import org.terasology.rendering.nui.widgets.types.TypeWidgetLibrary;
+import org.terasology.utilities.ReflectionUtil;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static org.terasology.rendering.nui.widgets.types.TypeWidgetFactory.LABEL_WIDGET_ID;
+
+public class FieldsWidgetBuilder<T> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FieldsWidgetBuilder.class);
+
+    private TypeInfo<T> type;
+    private TypeWidgetLibrary library;
+
+    private Map<Field, TypeInfo<?>> fields = Maps.newLinkedHashMap();
+
+    public FieldsWidgetBuilder(TypeInfo<T> type, TypeWidgetLibrary library) {
+        this(type, library, field -> true);
+    }
+
+    public FieldsWidgetBuilder(TypeInfo<T> type, TypeWidgetLibrary library,
+                               Predicate<Field> fieldPredicate) {
+        this.type = type;
+        this.library = library;
+
+        for (Field field : ReflectionUtils.getAllFields(type.getRawType())) {
+            if (Modifier.isTransient(field.getModifiers()) || Modifier.isStatic(field.getModifiers())) {
+                continue;
+            }
+
+            if (!fieldPredicate.test(field)) {
+                continue;
+            }
+
+            Type resolvedFieldType = ReflectionUtil.resolveType(type.getType(), field.getGenericType());
+
+            fields.put(field, TypeInfo.of(resolvedFieldType));
+        }
+    }
+
+    public static <F> Optional<Binding<F>> getFieldBinding(Binding<?> binding, Field field) {
+        if (Modifier.isPublic(field.getModifiers())) {
+            return Optional.of(
+                getAccessibleFieldBinding(binding, field)
+            );
+        }
+
+        return getPropertyBinding(binding, field);
+    }
+
+    private static <F> Optional<Binding<F>> getPropertyBinding(Binding<?> binding, Field field) {
+        Method setter = ReflectionUtil.findSetter(field);
+
+        if (setter == null) {
+            return Optional.empty();
+        }
+
+        Method getter = ReflectionUtil.findGetter(field);
+
+        if (getter == null) {
+            LOGGER.error("Cannot bind field {} with setter {} but no getter", field, setter);
+            return Optional.empty();
+        }
+
+        Binding<F> fieldBinding = new Binding<F>() {
+            @Override
+            public F get() {
+                try {
+                    return (F) getter.invoke(binding.get());
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Unreachable");
```

Comment:
Pretty ambiguous message, I'd also propagate `e` here and elsewhere in this class.